pywbem_mock/_mainprovider.py

```diff
@@ -0,0 +1,4208 @@
+            Indicates that class origin information is to be included on each
+            property in the returned instance, as follows:
+
+            * If `False`, class origin information is not included.
+            * If `True`, class origin information is included.
+            * If `None`, uses :term:`DSP0200` server default `False`.
+
+
+          PropertyList (:term:`string` or :term:`py:iterable` of :term:`string`):
+            An iterable specifying the names of the properties (or a string
+            that defines a single property) to be included in the returned
+            instance (case independent).
+
+            An empty iterable indicates to include no properties.
+
+            If `None`, all properties are included.
+
+        Returns:
+
+            A :class:`~pywbem.CIMInstance` object that is a representation of
+            the retrieved instance.
+            Its `path` attribute is a :class:`~pywbem.CIMInstanceName` object
+            with its attributes set as follows:
+
+            * `classname`: Name of the creation class of the instance.
+            * `keybindings`: Keybindings of the instance.
+            * `namespace`: Name of the CIM namespace containing the instance.
+            * `host`: `None`, indicating the WBEM server is unspecified.
+            TODO, host. Confirm what our mock server does.
+
+        Raises:
+
+            CIMError:  CIM_ERR_INVALID_NAMESPACE, CIM_ERR_INVALID_PARAMETER
+              CIM_ERR_NOT_FOUND
+        """  # noqa: E501
+        # pylint: enable:too-long
+
+        if InstanceName.namespace is None:
+            iname = InstanceName.copy()
+            iname.namespace = namespace
+        else:
+            iname = InstanceName
+
+        if iname.namespace != namespace:
+            raise CIMError(
+                CIM_ERR_INVALID_NAMESPACE,
+                _format("GetInstance Namespace {0!A} in InstanceName {1!A} "
+                        "does not match provided namespace.", iname.namespace,
+                        iname, namespace))
+
+        instance_store = self.get_instance_store(namespace)
+
+        if not self._class_exists(namespace, iname.classname):
+            raise CIMError(
+                CIM_ERR_INVALID_CLASS,
+                _format("Class {0!A} for GetInstance of instance {1!A} "
+                        "does not exist.", iname.classname, iname))
+
+        inst = self._get_instance(namespace, iname, instance_store,
+                                  PropertyList,
+                                  LocalOnly,
+                                  IncludeClassOrigin,
+                                  IncludeQualifiers)
+
+        return inst
+
+    ####################################################################
+    #
+    #   DeleteInstance server request
+    #
+    ####################################################################
+
+    def DeleteInstance(self, namespace, InstanceName):
```

Comment:
In DSP0200, the INSTANCENAME parameter does not specify a namespace, so it makes sense to have a separate target namespace. However, in this method, the namespace can be specified in the Instancename parameter and therefore is being checked in the implementation, leading to an exception about mismatching namespaces that would never even exist without that redundancy. I suggest to get rid of the redundancy by removing the `namespace` argument and by requiring that the namespace is set in the InstanceName argument.

Reply:
This one bothered me and in review of pegasus, it turns out we took exactly the opposite approach, keeping the namespace separate until the actual call to the physical repository. This allowed easier intermediate checks on the existence of the namespace itself but that was not a major issue.

Since at the point we call methods like mainprovider.DeleteInstance we have two possible sources of namespace, a namespace parameter which would correspond to the namespace element of the XML and is also provided by imethodcall OR a namespace  element in the path.

You prefer that the namespace element from the client (LOCALNAMESPACEPATH) be added to the InstanceName outside of the MainProvider.  I preferred that it be inserted into the instancename within the provider.

The error if the two namespaces do not match in minor. Can be removed by simply dictating that the namespace in InstanceName is ignored.  I just put it in to try to catch issues.

However, I felt that keeping the interface consistent so that there was a namespace element for every call was what I wanted in consistency. You prefer to specialize these 3 mainprovider methods slightly by making the caller insure that the namespace is within the InstancName.

We are probably down to minor semantics here, common parameters across all of the mainprovider responder methods, or these 3 slightly special because they contain InstanceName which can carry namespace within it. I picked the separation because it corresponds to how the XML carries namespace (however, it means that the 3 methods move the namespace into the instance, you pick integration because it simplifies the interface.

NOTE: My error could only be tested by creating new tests at the MainProvider level.

**DISCUSSION**

Reply:
"in review of pegasus, it turns out we took exactly the opposite approach, keeping the namespace separate until the actual call to the physical repository": The pywbem_mock provider is very close to the repository, and it is always using a repository.

I just don't see the value of ignoring the namespace on pywbem path objects like CIMinstanceName or CIMClassName when in other places we even ensure that they have their namespace set.

But let's discuss.

Reply:
I agree and will remove the namespace on GetInstance, DeleteInstance, and ModifyInstance.

Reply:
done